File: kotlin/src/main/kotlin/exercises/FirstWorld.kt
package exercises

import display.Color
import display.Resolution
import display.canvasToPpm
import display.writePpm
import geometry.Point
import geometry.Vector
import geometry.WORLD_ORIGIN
import geometry.rotationX
import geometry.rotationY
import geometry.scaling
import geometry.translation
import geometry.viewTransform
import tracing.Camera
import tracing.Material
import tracing.PointLight
import tracing.Sphere
import tracing.World
import kotlin.system.measureTimeMillis

fun main() {

    val runtime = measureTimeMillis {

        val wallMaterial = Material(Color(1, 0.9, 0.9), specular = 0.0)

        val floor = Sphere()
            .withOrigin(WORLD_ORIGIN)
            .withTransform { scaling(10, 0.01, 10) }
            .withMaterial(wallMaterial)

        val leftWall = Sphere()
            .withOrigin(WORLD_ORIGIN)
            .withTransform {
                scaling(10, 0.01, 10)
                    .then(rotationX(Math.PI / 2))
                    .then(rotationY(-Math.PI / 4))
                    .then(translation(0, 0, 5))
            }
            .withMaterial(wallMaterial)

        val rightWall = Sphere()
            .withOrigin(WORLD_ORIGIN)
            .withTransform {
                scaling(10, 0.01, 10)
                    .then(rotationX(Math.PI / 2))
                    .then(rotationY(Math.PI / 4))
                    .then(translation(0, 0, 5))
            }
            .withMaterial(wallMaterial)

        val middleTransform = translation(-0.5, 1, 0.5)
        val middleMaterial = Material(Color(0.1, 1, 0.5), diffuse = 0.7, specular = 0.3)
        val middle = Sphere(middleTransform, WORLD_ORIGIN, middleMaterial)

        val rightTransform = scaling(0.5, 0.5, 0.5).then(translation(1.5, 0.5, -0.5))
        val right = Sphere(rightTransform, WORLD_ORIGIN, middleMaterial)

        val leftTransform = scaling(0.33, 0.33, 0.33).then(translation(-1.5, 0.33, -0.75))
        val leftMaterial = Material(Color(1, 0.8, 0.1), diffuse = 0.7, specular = 0.3)
        val left = Sphere(leftTransform, WORLD_ORIGIN, leftMaterial)

        val lightSource = PointLight(Point(-10, 10, -10), Color(1, 1, 1))
        val world = World(listOf(floor, leftWall, rightWall, middle, right, left), lightSource)

        val cameraTransform = viewTransform(Point(0, 1.5, -5), Point(0, 1, 0), Vector(0, 1, 0))
        val camera = Camera(Resolution(400, 200), Math.PI / 3, cameraTransform)

        val canvas = camera.render(world)
        writePpm(canvasToPpm(canvas))
    }

    println("Image created in $runtime")
}

File: rust/src/exercises/first_world.rs
use crate::display::color::Color;
use crate::display::writer::write_canvas;
use crate::geometry::point::Point;
use crate::geometry::transformations::{
    rotation_x, rotation_y, scaling, translation, view_transform,
};
use crate::geometry::vector::Vector;
use crate::tracing::camera::Camera;
use crate::tracing::material::Material;
use crate::tracing::point_light::PointLight;
use crate::tracing::sphere::Sphere;
use crate::tracing::world::World;
use std::f64::consts::PI;

pub fn run() {
    let wall_material = Material::default()
        .with_color(Color::new(1.0, 0.9, 0.9))
        .with_specular(0.0);

    let floor = Sphere::new(
        Point::origin(),
        wall_material.clone(),
        scaling(10.0, 0.01, 10.0),
    );

    let transform = scaling(10.0, 0.01, 10.0)
        .then(&rotation_x(PI / 2.0))
        .then(&rotation_y(-PI / 4.0))
        .then(&translation(0, 0, 5));

    let left_wall = Sphere::new(Point::origin(), wall_material.clone(), transform.clone());

    let right_wall = Sphere::new(Point::origin(), wall_material.clone(), transform.clone());

    let middle_transform = translation(-0.5, 1.0, 0.5);
    let middle_material = Material::default()
        .with_color(Color::new(0.1, 1.0, 0.5))
        .with_diffuse(0.7)
        .with_specular(0.3);

    let middle = Sphere::new(Point::origin(), middle_material.clone(), middle_transform);

    let right_transform = scaling(0.5, 0.5, 0.5).then(&translation(1.5, 0.5, -0.5));

    let right = Sphere::new(Point::origin(), middle_material, right_transform);

    let left_transform = scaling(0.33, 0.33, 0.33).then(&translation(-1.5, 0.33, -0.75));

    let left_material = Material::default()
        .with_color(Color::new(1.0, 0.8, 0.1))
        .with_diffuse(0.7)
        .with_specular(0.3);

    let left = Sphere::new(Point::origin(), left_material, left_transform);

    let light_source = PointLight::new(Point::at(-10, 10, -10), Color::new(1, 1, 1));
    let world = World::new(vec![floor, left_wall, right_wall, middle, right, left], light_source,);

    let camera_transform = view_transform(
        Point::at(0.0, 1.5, -5.0),
        Point::at(0, 1, 0),
        Vector::new(0, 1, 0),
    );
    let camera = Camera::new(400, 200, PI / 3.0, camera_transform);

    let canvas = camera.render(world);
    write_canvas(&canvas);
}
